lib/Dialect/Handshake/HandshakeTypes.cpp

```diff
@@ -11,55 +11,59 @@
 //===----------------------------------------------------------------------===//
 
 #include "dynamatic/Dialect/Handshake/HandshakeTypes.h"
+#include "dynamatic/Support/LLVM.h"
 #include "mlir/IR/Builders.h"
 #include "mlir/IR/Diagnostics.h"
 #include "mlir/IR/DialectImplementation.h"
 #include "mlir/IR/OpImplementation.h"
 #include "mlir/Support/LLVM.h"
 #include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/ADT/TypeSwitch.h"
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/JSON.h"
 #include <cctype>
+#include <iostream>
+#include <ostream>
 
 using namespace mlir;
 using namespace dynamatic;
 using namespace dynamatic::handshake;
 
 //===----------------------------------------------------------------------===//
-// ChannelType
+// Common implementation for ControlType and ChannelType
 //===----------------------------------------------------------------------===//
 
-unsigned dynamatic::handshake::getHandshakeTypeBitWidth(Type type) {
-  return llvm::TypeSwitch<Type, unsigned>(type)
-      .Case<handshake::ControlType>([](auto) { return 0; })
-      .Case<handshake::ChannelType>(
-          [](ChannelType channelType) { return channelType.getDataBitWidth(); })
-      .Case<IntegerType, FloatType>(
-          [](Type type) { return type.getIntOrFloatBitWidth(); })
-      .Default([](Type type) {
-        llvm_unreachable("unsupported type");
-        return 0;
-      });
-}
-
 static constexpr llvm::StringLiteral UPSTREAM_SYMBOL("U");
 
-/// Checks the validity of a channel's data type.
-static LogicalResult
-checkChannelData(function_ref<InFlightDiagnostic()> emitError, Type dataType) {
-  if (!ChannelType::isSupportedSignalType(dataType)) {
-    return emitError()
-           << "expected data type to be IntegerType or FloatType, but got "
-           << dataType;
-  }
-  if (dataType.getIntOrFloatBitWidth() == 0) {
-    return emitError()
-           << "expected data type to have strictly positive bitwidth, but got "
-           << dataType;
+/// Print the extra signals to generate an IR representation.
+/// For example: [spec: i1, tag: i32 (U)]
+/// This function is common to both ControlType and ChannelType.
+static void printExtraSignals(AsmPrinter &odsPrinter,
+                              llvm::ArrayRef<ExtraSignal> extraSignals) {
+
+  // Print a single signal.
+  // eg. spec: i1
+  auto printSignal = [&](const ExtraSignal &signal) {
+    odsPrinter << signal.name << ": " << signal.type;
+    if (!signal.downstream)
+      odsPrinter << " (" << UPSTREAM_SYMBOL << ")";
+  };
+
+  // Print all signals enclosed in square brackets
+  odsPrinter << "[";
+  for (const ExtraSignal &signal : extraSignals.drop_back()) {
```

Comment:
This will crash if there are 0 extra signals. If the function assumes it cannot happen it needs to be documented and have an accompanying assert

Reply:
Thank you for pointing this out. I added both the documentation comment and the assertion